### PR TITLE
LogExecution: optionally take a module name for the logger function

### PR DESCRIPTION
**Pull request by yamt**
also, fix a crash on modules w/o imports.

### CODE REVIEW

**Comment by kripken** [comment]
I agree that the automatic behavior here can be hard to use. But we do need a solution for the case where `"env"` was minified, which that PR solved. How about making the automatic behavior customizable? We could perhaps allow not just `--log-execution` but also `--log-execution=PREFIX`, and in the second case that exact prefix would be used and nothing else (no automatic behavior). Would that work for your use case @yamt ?

If so then to implement it see e.g. `ExtractFunction.cpp` for an example of a pass that receives a parameter like that.

**Comment by yamt** [comment]
sure, making it an option works for me.
however, i don't understand why the automatic behavior is necessary in the first place.
probably because i'm not familiar with how minimization works.
can't you simply minimize after applying log-execute instrumentation?

**Comment by kripken** [comment]
You're right that a different order might work. But sometimes you want to add the logging after all optimizations have been done, maybe because the bug only shows itself then (maybe it's a bug in an optimization pass).

In general I've personally found this pass useful on various wasm files, optimized and not, and I'm not sure how other people are using the pass, so I'd rather not remove functionality.

**Comment by yamt** [comment]
> You're right that a different order might work. But sometimes you want to add the logging after all optimizations have been done, maybe because the bug only shows itself then (maybe it's a bug in an optimization pass).
> 
> In general I've personally found this pass useful on various wasm files, optimized and not, and I'm not sure how other people are using the pass, so I'd rather not remove functionality.

ok. it makes sense.

**Comment by yamt** [comment]
i added the suggested option.

File: src/passes/LogExecution.cpp [hunk 1]
```diff
@@ -39,9 +39,17 @@ namespace wasm {
 Name LOGGER("log_execution");
 
 struct LogExecution : public WalkerPass<PostWalker<LogExecution>> {
+  IString logger_module;
```
Comment (kripken) [inline src/passes/LogExecution.cpp:42]:
```suggestion
  // The module name the logger function is imported from.
  Name loggerModule;
```

File: src/passes/LogExecution.cpp [hunk 2]
```diff
@@ -39,9 +39,17 @@ namespace wasm {
...
+
   // Adds calls to new imports.
   bool addsEffects() override { return true; }
 
+  void run(Module* module) override {
+    auto& options = getPassOptions();
```
Comment (kripken) [inline src/passes/LogExecution.cpp:48]:
```suggestion
```

File: src/passes/LogExecution.cpp [hunk 3]
```diff
@@ -63,23 +71,31 @@ struct LogExecution : public WalkerPass<PostWalker<LogExecution>> {
...
+      if (!import->module) {
+        for (auto& func : curr->functions) {
+          if (func->imported()) {
+            import->module = func->module;
+            break;
+          }
+        }
+      }
+
+      if (!import->module) {
```
Comment (kripken) [inline src/passes/LogExecution.cpp:96]:
```suggestion
      // If no function was found, use ENV.
      if (!import->module) {
```

**Review by kripken — COMMENTED** [review]
Please add a test for this. It could be similar to this test file that I just opened a PR for,

https://github.com/WebAssembly/binaryen/pull/6634/files#diff-26883f6202548dd332237e9002be15d8e33b958bdbbb52eef03e0999e147dd6e

Maybe call it `test/lit/passes/log-execution_arg.wast` (adding `_arg`), and put a comment in it saying
```
;; Test the option to provide the module name as an argument.
```
Then you can just modify the `wasm-opt` command there. To update the expected output there use
```
./scripts/update_lit_checks.py test/lit/passes/log-execution_arg.wast
```

File: src/passes/LogExecution.cpp (hunk 2 above)
Comment (yamt) [reply src/passes/LogExecution.cpp:48]: i don't understand this one. can you explain a bit?

**Comment by yamt** [comment]
> Please add a test for this. It could be similar to this test file that I just opened a PR for,
> 
> https://github.com/WebAssembly/binaryen/pull/6634/files#diff-26883f6202548dd332237e9002be15d8e33b958bdbbb52eef03e0999e147dd6e
> 
> Maybe call it `test/lit/passes/log-execution_arg.wast` (adding `_arg`), and put a comment in it saying
> 
> ```
> ;; Test the option to provide the module name as an argument.
> ```
> 
> Then you can just modify the `wasm-opt` command there. To update the expected output there use
> 
> ```
> ./scripts/update_lit_checks.py test/lit/passes/log-execution_arg.wast
> ```

done.
i needed `--all-items` to make it check the import.

File: src/passes/LogExecution.cpp (hunk 1 above)
Comment (yamt) [reply src/passes/LogExecution.cpp:43]: done

File: src/passes/LogExecution.cpp (hunk 3 above)
Comment (yamt) [reply src/passes/LogExecution.cpp:98]: done

File: src/passes/LogExecution.cpp (hunk 2 above)
Comment (kripken) [reply src/passes/LogExecution.cpp:49]: Sorry, I misread the code here, this suggestion of mine was wrong.

File: src/passes/LogExecution.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/LogExecution.cpp:43]: Aside from the comment, the diff also renamed the variable. `loggerModule` is more consistent with our naming conventions (compared to `logger_module`).

File: test/lit/passes/log-execution_arg.wast [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  ;; CHECK:      (func $nopp
+  ;; CHECK-NEXT:  (call $log_execution
+  ;; CHECK-NEXT:   (i32.const 0)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (nop)
+  ;; CHECK-NEXT: )
+  (func $nopp
+    (nop)
+  )
+  ;; CHECK:      (func $intt (result i32)
```
Comment (kripken) [inline test/lit/passes/log-execution_arg.wast:25]: I think it's ok to stop this test here. The first 25 lines show that it uses "foo" and that it emits a valid call to `log_execution`.

**Review by kripken — COMMENTED** [review]
Looks good aside from those last comments.

File: test/lit/passes/log-execution_arg.wast (hunk 4 above)
Comment (kripken) [reply test/lit/passes/log-execution_arg.wast:25]: (That is, it doesn't need to be as comprehensive as the other test, as the focus here is just on the argument to the pass, which lines 1-25 test.)

Comment (yamt) [reply test/lit/passes/log-execution_arg.wast:25]: are you suggesting to minimize the test?

File: src/passes/LogExecution.cpp (hunk 1 above)
Comment (yamt) [reply src/passes/LogExecution.cpp]: oops, i haven't noticed it. done.

File: test/lit/passes/log-execution_arg.wast (hunk 4 above)
Comment (yamt) [reply test/lit/passes/log-execution_arg.wast]: done

Comment (kripken) [reply test/lit/passes/log-execution_arg.wast]: Yes, exactly.

**Review by kripken — APPROVED** [review]

**Pull request merged**